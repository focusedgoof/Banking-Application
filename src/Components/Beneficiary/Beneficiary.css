.beneficiary {
  padding: 20px;
  display: flex;
  flex-wrap: wrap;
  min-height:84.5vh;
}

.beneficiary > h3 {
  display: flex;
  font-size: 2rem;
  width: 100%;
  font-weight: bold;
  justify-content: center;
  margin-bottom:3%;
  margin-top:0;
}

.beneficiary .common-table{
  width:90%;
}
.empty-table{
  width:100%;
  font-size: 1.2rem;
  font-weight:bold;
  color:#34b3f7;
  margin-bottom:4%;

}

.add-beneficiary{  
  width:50%;
  display:flex;
  margin:auto;
  flex-direction: column;
  margin-top:4%;
  background-color:#f0f0f0;
  border-radius:5px;
  margin-bottom:2%;
}
.add-beneficiary .form-container {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  width: 100%;
  margin: 20px auto;
}

.add-beneficiary h3{
  font-weight:bold;
  font-size: 1.8rem;
  color:#03658C;
}
.beneficiary .form-container{
  width:60%;  
}
.beneficiary .form-container button{
  padding: 3%;
}



