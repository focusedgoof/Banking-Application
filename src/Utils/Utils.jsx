export async function getLocalStorage(key) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        const storedValue = localStorage.getItem(key);
        resolve(storedValue ? JSON.parse(storedValue) : null);
      } catch (error) {
        reject(error);
      }
    }, 0);
  });
}

export async function setLocalStorage(key, value) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      try {
        localStorage.setItem(key, JSON.stringify(value));
        resolve();
      } catch (error) {
        reject(error);
      }
    }, 0);
  });
}
