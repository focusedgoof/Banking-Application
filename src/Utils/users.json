[
  {
    "Username": "Akshat Arya",
    "Password": "Passw0rd",
    "AccountNumber": 1234567890123456,
    "Balance": 6707,
    "PAN": "ABCD1234E",
    "Branch": "Branch A",
    "IFSCCode": "IFSC001"
  },
  {
    "Username": "Rahul Sharma",
    "Password": "Secur1ty",
    "AccountNumber": 9876543210987654,
    "Balance": 5635,
    "PAN": "WXYZ5678F",
    "Branch": "Branch B",
    "IFSCCode": "IFSC002"
  },
  {
    "Username": "Umang Mittal",
    "Password": "Log1n123",
    "AccountNumber": 4567890123456789,
    "Balance": 60375,
    "PAN": "PQRS9012G",
    "Branch": "Branch C",
    "IFSCCode": "IFSC003"
  },
  {
    "Username": "Priya Gupta",
    "Password": "NewPassw0rd",
    "AccountNumber": 7890123456789012,
    "Balance": 10000,
    "PAN": "EFGH5678H",
    "Branch": "Mumbai Central",
    "IFSCCode": "IFSC004"
  }
]
  