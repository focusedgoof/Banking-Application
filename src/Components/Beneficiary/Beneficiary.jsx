import React, { useState, useEffect, useRef } from 'react';
import { setLocalStorage,getLocalStorage } from '../../Utils/Utils';
import SnackBar from '../Common/SnackBar/SnackBar';
import Form from '../Common/Form/Form';  
import CommonTable from '../Common/Table/CommonTable';
import './Beneficiary.css';

const Beneficiary = () => {
  const [userBeneficiary, setUserBeneficiary] = useState([]);  
  const [filteredBeneficiary, setFilteredBeneficiary] = useState([]);
  const [searchText, setSearchText] = useState('');
  const [page, setPage] = useState(0);
  const formRef = useRef(null);  
  const columns = [
    { key: 'Beneficiary', label: "Beneficiary Name" },
    { key: 'AccountBeneficiary', label: 'Account Number' },
    { key: 'BranchBeneficiary', label: 'Branch Name' },
  ];
  const [snackbarInfo, setSnackbarInfo] = useState({
    open: false,
    message: '',
    severity: 'error',
  });

  useEffect(() => {
    const fetchData = async () => {
      try {
        const userDataString = await getLocalStorage('userData');
        if (userDataString) {
          try {
            let storedBeneficiaryData = await getLocalStorage('BeneficiaryData');
  
            if (!storedBeneficiaryData) {
              storedBeneficiaryData = [];
            }
  
            const userSpecificBeneficiary = storedBeneficiaryData.filter(
              (data) => data.User === userDataString.Username
            );
  
            setUserBeneficiary(userSpecificBeneficiary);
          } catch (parseError) {
            console.error('Error parsing user data:', parseError);
          }
        }
      } catch (error) {
        console.error('Error fetching user data from local storage:', error);
      }
    };
  
    fetchData();
  }, []);
  

  useEffect(() => {
    setFilteredBeneficiary(userBeneficiary);
  }, [userBeneficiary]);

  const isValidAccountNumber = (accNumber) => /^\d+$/.test(accNumber);
  const containsDigit = (text) => /\d/.test(text);

  const handleAddBeneficiary = async (formData) => {
    const { AccountBeneficiary, Beneficiary, BranchBeneficiary } = formData;
  
    if (containsDigit(Beneficiary) || 
        containsDigit(BranchBeneficiary) ||
        !isValidAccountNumber(AccountBeneficiary)) {
      setSnackbarInfo({
        open: true,
        message: 'Please fill in all the required fields',
      });
      return;
    }
  
    try {
      const userDataString = await getLocalStorage('userData');
      if (userDataString) {
        try {
          let storedBeneficiaryData = await getLocalStorage('BeneficiaryData');
  
          if (!storedBeneficiaryData) {
            storedBeneficiaryData = [];
          }
  
          const newBeneficiaryWithUserDetails = {
            User: userDataString.Username,
            AccountUser: userDataString.AccountNumber,
            Beneficiary,
            AccountBeneficiary,
            BranchBeneficiary,
          };
  
          storedBeneficiaryData.push(newBeneficiaryWithUserDetails);
  
          await setLocalStorage('BeneficiaryData', storedBeneficiaryData);
  
          setUserBeneficiary((prevBeneficiaries) => {
            return [...prevBeneficiaries, newBeneficiaryWithUserDetails];
          });
  
          setSnackbarInfo({
            open: true,
            message: 'Beneficiary added successfully!',
            severity: 'success',
          });
        } catch (parseError) {
          console.error('Error parsing user data:', parseError);
        }
      }
    } catch (error) {
      console.error('Error storing beneficiary data:', error);
    }
  
    if (formRef.current) {
      formRef.current.reset();
    }
  };
  

  const handleSnackbarClose = () => {
    setSnackbarInfo({ ...snackbarInfo, open: false });
  };

  const handleSearch = (value) => {
    setSearchText(value);

    const filtered = userBeneficiary.filter((beneficiary) => {
      const values = Object.values(beneficiary).join(" ").toLowerCase();
      return values.includes(value.toLowerCase());
    });
    setFilteredBeneficiary(filtered);
    setPage(0);
  };

  return (
    <section className="beneficiary">
      <h3>Beneficiary Details</h3>
      {userBeneficiary.length > 0 ? (
        <div className="common-table">
          <CommonTable
            columns={columns}
            data={filteredBeneficiary}
            searchText={searchText}
            page={page}
            handleSearch={handleSearch}
          />
        </div>
      ) : (
        <p className="empty-table">No existing beneficiary. Add a beneficiary from the form below.</p>
      )}
      <div className="add-beneficiary">
        <h3>Add new beneficiary</h3>
        <div className="form-container">
          <Form
            formFields={[
              { name: 'Beneficiary', placeholder: "Beneficiary Name" },
              { name: 'AccountBeneficiary', placeholder: 'Account Number' },
              { name: 'BranchBeneficiary', placeholder: 'Branch Name' },
            ]}
            onSubmit={handleAddBeneficiary}
            buttonLabel="Add Beneficiary"
            ref={formRef}
          />
        </div>
      </div>

      <SnackBar {...snackbarInfo} handleClose={handleSnackbarClose} />
    </section>
  );
};

export default Beneficiary;
