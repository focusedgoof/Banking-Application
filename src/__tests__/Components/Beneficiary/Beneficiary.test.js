import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Beneficiary from '../../../Components/Beneficiary/Beneficiary';

jest.mock('react', () => {
  const actualReact = jest.requireActual('react');
  return {
    ...actualReact,
    useRef: jest.fn(() => ({ current: {} })),
  };
});

// Mocking localStorage functions
jest.mock('../../../Utils/Utils', () => ({
  getLocalStorage: jest.fn(),
  setLocalStorage: jest.fn(),
}));

// Mock the fetchData function inside useEffect
jest.mock('../../../Components/Beneficiary/Beneficiary', () => {
  return {
    __esModule: true,
    default: () => <div>Mocked Beneficiary Component</div>,
  };
});

describe('Beneficiary Component Snapshot Test', () => {
  test('renders Beneficiary component correctly with existing beneficiaries', () => {
    const existingBeneficiaries = [
      {
        User: 'testUser',
        AccountUser: '1234567890',
        Beneficiary: 'John Doe',
        AccountBeneficiary: '0987654321',
        BranchBeneficiary: 'Test Branch',
      },
      // Add more beneficiary objects as needed
    ];

    const getLocalStorageMock = jest.fn();
    getLocalStorageMock.mockReturnValueOnce(JSON.stringify(existingBeneficiaries));
    require('../../../Utils/Utils').getLocalStorage = getLocalStorageMock;

    const { asFragment } = render(<Beneficiary />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Beneficiary component correctly with no existing beneficiaries', () => {
    const getLocalStorageMock = jest.fn();
    getLocalStorageMock.mockReturnValueOnce(JSON.stringify([]));
    require('../../../Utils/Utils').getLocalStorage = getLocalStorageMock;

    const { asFragment } = render(<Beneficiary />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Beneficiary component correctly', () => {
    const getLocalStorageMock = jest.fn();
    getLocalStorageMock.mockReturnValueOnce(JSON.stringify([]));
    require('../../../Utils/Utils').getLocalStorage = getLocalStorageMock;

    const { asFragment } = render(<Beneficiary />);
    expect(asFragment()).toMatchSnapshot();
  });
});
