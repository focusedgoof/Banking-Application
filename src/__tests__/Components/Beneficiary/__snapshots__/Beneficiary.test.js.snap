// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beneficiary Component Snapshot Test renders Beneficiary component correctly 1`] = `
<DocumentFragment>
  <div>
    Mocked Beneficiary Component
  </div>
</DocumentFragment>
`;

exports[`Beneficiary Component Snapshot Test renders Beneficiary component correctly with existing beneficiaries 1`] = `
<DocumentFragment>
  <div>
    Mocked Beneficiary Component
  </div>
</DocumentFragment>
`;

exports[`Beneficiary Component Snapshot Test renders Beneficiary component correctly with no existing beneficiaries 1`] = `
<DocumentFragment>
  <div>
    Mocked Beneficiary Component
  </div>
</DocumentFragment>
`;
